### ISSUE TITLE
TreeView OverflowException

### ISSUE THREAD

**Issue by coderb** [dotnet/winforms#7098]
### .NET version

net6 x86

### Did it work in .NET Framework?

Yes

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

unknown

### Issue description

I'm hitting an OverflowException in TreeView.cs due to the cast to IntPtr on x86 (32-bit) with multiple monitors.

It looks like this is likely fixed under net7 with https://github.com/dotnet/winforms/pull/5791 .

`User32.SendMessageW(this, User32.WM.CONTEXTMENU, Handle, (IntPtr)User32.GetMessagePos());`

The problem is that `GetMessagePos()` could return 0xFFFEFDB5 which overflows on 32-bit
```
            //var x = GetMessagePos(); // -> 0xFFFEFDB5
            uint n = 0xFFFEFDB5;
            var nn = (IntPtr)n; // throws OverflowException
```

Is it possible to get a fix for this in net6 since net7 is still 1/2 year+ away?

### Steps to reproduce

see above

**Comment by dreddy-work** [comment]
Thanks @coderb for reporting this. @coderb do you mind checking if there are any other instances like this in the WinForms code base? That would meet the bar for servicing and we will work on the fix.

**Comment by coderb** [comment]
@dreddy-work that's the only instance of an overflow exception that i hit.  unfortunately, the codebase probably has many issues (see the merged patches for net7 and discussion here ->  https://github.com/dotnet/winforms/pull/5791 ).

**Comment by dreddy-work** [comment]
this require majority of #5791 to be ported not just tailored fix here. @JeremyKuhne any thoughts on this?

**Comment by JeremyKuhne** [comment]
@dreddy-work We wouldn't port the whole of #5791 for a servicing fix. A targeted fix can be evaluated for servicing. @coderb, please confirm that this is the location you're hitting the issue.

https://github.com/dotnet/winforms/blob/ec6689c3ebccb90aeea7c6912648da348588018a/src/System.Windows.Forms/src/System/Windows/Forms/TreeView.cs#L3098

The current implementation is this:

https://github.com/dotnet/winforms/blob/5ccc9c034b6f792d3d03cc28d12766e0833981af/src/System.Windows.Forms/src/System/Windows/Forms/TreeView.cs#L3147

The targeted fix is this:

``` C#
User32.SendMessageW(this, User32.WM.CONTEXTMENU, Handle, (IntPtr)(int)User32.GetMessagePos());
```


Note that this didn't happen on .NET Framework as `GetMessagePos` was defined to return `int` (technically incorrectly as it is a `DWORD`).

https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/TreeView.cs,3048

**Comment by coderb** [comment]
@JeremyKuhne yes, that is the line that has the probmlem.

**Comment by dreddy-work** [comment]
Thanks @JeremyKuhne , i was thinking in the lines of making nint changes across the call flow here . Hence, realized it needed most those changes in #5791. Linked [PR](https://github.com/dotnet/winforms/pull/7183) has targeted fix. Will be sending it to servicing approval.

**Comment by MelonWang1** [comment]
@dreddy-work I can't repro this issue in .NET 6.0.5 or .NET 5.0, maybe is my repro steps are incorrect, could you please help me confirm this issue?
**Repro scenario1:**
1. Create a winforms .net core 6.0 project(both 6.0.5 and 6.0.6)
2. Change the Platform target to x86 and add a treeview control to form designer
3. Build and run, click treeview control

**Result1:** No Overflow Exception occurs.

**Repro scenario2:**
1. Create a winforms .net core 6.0 project(both 6.0.5 and 6.0.6)
2. Change the Platform target to x86 and add a button control
3. Double click button control and add blow code
            //var x = GetMessagePos(); // -> 0xFFFEFDB5
            uint n = 0xFFFEFDB5;
            var nn = (IntPtr)n; // throws OverflowException
5. Build and run, click button control

**Result2:** The Overflow Exception throws.

@coderb If my steps are incorrect, could you please give me a sample project to repro and verify this issue?

**Comment by coderb** [comment]
@MelonWang1 melon wang-> you also need to have multiple monitors configured in such a way that one of them has negative coordinates and then move the control into that area of the display.

**Comment by MelonWang1** [comment]
@coderb I set the negative coordinates of monitors, but I still not repro this issue in .NET core 5.0, below is my repro scenario, could you please help me check my steps? if I am missing some steps please let me know.
**Repro scenario:**
1. Set the second monitor is left of main monitor, the second monitor has negative coordinates
2. Create a winforms .net core 5.0 project
3. Right click the project and select Properties -> Build ->Platform target, change Any CUP to x86
4. Drag and drop a treeview control to main monitors and then move it to second monitor
4. Build and run, click treeview control
**Result:** No Overflow Exception occurs.

**Comment by coderb** [comment]
@MelonWang1 i'm not sure exactly when `(IntPtr)n` generates an OverflowException, but for my particular monitor setup I was able to capture the exception in the debugger and the `x` coordinate was `0xFFEFDB5`.  

Perhaps you can log the result of GetMessagePos() and try to move the app around to attempt to get it to be as close to 0xFFFFFFFF as possible.

Also the exception was happening when the checkbox of the treeview control is clicked.   Thanks.

**Comment by MelonWang1** [comment]
We are simulating the product code to verify this issue, in .NET SDK 6.0 TP build, the issue was fixed.
**Verify steps:**
	1. Create a winforms .net core 6.0 project
	2. Change the Platform target to x86 and add a button control
	3. Double click button control and add below code, then build and run, click the button
**Result:**
when the code is:
            //var x = GetMessagePos(); // -> 0xFFFEFDB5
            uint n = 0xFFFEFDB5;
            var nn = (IntPtr)n; // throws OverflowException
The Overflow Exception throws.
![7098bad](https://user-images.githubusercontent.com/94418985/171336360-9056c610-ddd3-445e-8117-359b58d7cdc9.gif)

**When the code is simulating the product code:**
            //var x = GetMessagePos(); // -> 0xFFFEFDB5
            uint n = 0xFFFEFDB5;
            var nn = (IntPtr)(int)n; // throws OverflowException
**No Overflow Exception occurs.**
![7098good](https://user-images.githubusercontent.com/94418985/171336386-fa4be1d0-cf37-4d2e-9a40-b4fa206b7769.gif)
@coderb Thank you for helping us to reproduce this issue, but we still couldn't reproduce the issue with the tree view control and negative coordinates, so we ended up using a button control to verify the issue by simulating the product code.
